File: 0789-kth-largest-element-in-a-stream/0789-kth-largest-element-in-a-stream.py
import heapq
from typing import List

class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        """
        Initialize the object with the integer k and the list of integers nums.
        """
        self.k = k
        self.min_heap = []
        
        # Add the initial numbers to the heap
        for num in nums:
            self.add(num)
    
    def add(self, val: int) -> int:
        """
        Add a new integer val to the data structure and return the k-th largest element.
        """
        if len(self.min_heap) < self.k:
            heapq.heappush(self.min_heap, val)
        else:
            # Only add the new value if it's larger than the smallest in the heap
            if val > self.min_heap[0]:
                heapq.heapreplace(self.min_heap, val)
        
        # The k-th largest element is the smallest element in the heap
        return self.min_heap[0]

# Example usage:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: 0789-kth-largest-element-in-a-stream/0789-kth-largest-element-in-a-stream.cpp
#include <vector>
#include <queue>
using namespace std;

class KthLargest {
public:
    KthLargest(int k, vector<int>& nums) : k(k) {
        // Use a min-heap to store the k largest elements
        for (int num : nums) {
            add(num);
        }
    }
    
    int add(int val) {
        // If the heap has fewer than k elements, push the new value
        if (minHeap.size() < k) {
            minHeap.push(val);
        } else if (val > minHeap.top()) {
            // If the heap has k elements and the new value is larger than the smallest in the heap
            minHeap.pop();
            minHeap.push(val);
        }
        // The top of the min-heap is the k-th largest element
        return minHeap.top();
    }

private:
    int k;
    priority_queue<int, vector<int>, greater<int>> minHeap; // Min-heap to store k largest elements
};
